draft-pda-core.md

```diff
@@ -291,11 +293,427 @@ Each PA task in this document is divided into three sub-protocols as follows.
+GET request to `[helper]/key_config`. The helper responds with status 200 and an
+`HpkeConfig` message. Next, the client collects the set of helpers it will
+upload shares to. It ignores a helper if:
+* the client and helper failed to establish a secure, helper-authenticated
+  channel;
+* the GET request to the helper URL failed or didn't return a valid key config;
+  or
+* the key config specifies a KEM, KDF, or AEAD algorithm the client doesn't
+  recognize.
+
+If the set of supported helpers is empty, then the client aborts and alerts the
+leader with "no supported helpers". Otherwise, for each supported helper the
+client issues a POST request to `[leader]/upload_finish` with a payload
+constructed as described below.
+
+[OPEN ISSUE: @chris-wood: Can't the leader determine if helpers are "online"?
+This seems to reveal information that's specific to clients. Imagine, for
+example, that clients are prohibited from talking to helpers but not the leader.
+Is it OK that leaders learn that about a client? I'm not sure, so I'd be
+inclined to remove this unless we have a concrete use case.]
+
+The client begins by setting up an HPKE {{!I-D.irtf-cfrg-hpke}} context for
+the helper by running
+
+~~~
+enc, context = SetupBaseS(pk, [TODO])
+~~~
+
+where `pk` is the KEM public key encoded by `HpkeConfig.public_key`. The outputs
+are the helper's encapsulated context `enc` and the context `context`.
+
+[TODO: Decide what the info string should be. At a minimum it should include
+`PAParam` so that the PA parameters are implicitly authenticated by the helper.
+when it decapsulates `enc`. We might consider using the hash of the transcript
+between the client and leader so far, i.e., `PAUploadStartReq`, ...,
+`PAUploadStartResp`, as this would be the most conservative thing. We just need
+to ensure that the helper can reproduce this transcript. This should be possible
+given the information it has.]
+
+Next, the client encodes its measurements as an input for the PA protocol. It
+then generates a validity proof for its input and uses `context` to split the
+input and proof into a *leader share* and a *helper share*, where the latter is
+protected by the HPKE context. Note that the details of each of these processing
+steps --- encode, prove, split, and encrypt --- are specific to the PA protocol.
+
+[OPEN ISSUE: Is it safe to generate the proof once, then secret-share between
+each (leader, helper) pair? Probably not in general, but maybe for Prio?]
+
+[OPEN ISSUE: allow server to send joint randomness in UploadStartResp, and then
+enforce uniqueness via double-spend state or something else]
+
+The payload of the POST request to `[leader]/upload_finish` is structured as
+follows:
+
+~~~
+struct {
+  PATask task;                 // Equal to PAUploadStartReq.task
+  uint8 helper_hpke_config_id; // Equal to HpkeConfig.id
+  Url helper_url;
+  PAHelperShare helper_share;
+  PALeaderShare leader_share;
+} PAUploadFinishReq;
```

Comment:
Why are we replicating the leader_share for each helper?

Reply:
There's only one helper here, so no duplication.

Reply:
"Otherwise, for each supported helper the client issues a POST request to `[leader]/upload_finish` with a payload
constructed as described below.

Reply:
Oh I see. Will flag as an open issue.